src/core/device.c

```diff
@@ -189,41 +182,64 @@ static int device_coldplug(Unit *u) {
          *
          * - On switch-root, the udev database may be cleared, except for devices with sticky bit, i.e.
          *   OPTIONS="db_persist". Hence, almost no devices are enumerated in the step 2. However, in general,
-         *   we have several serialized devices. So, DEVICE_FOUND_UDEV bit in the deserialized_found must be
-         *   ignored, as udev rules in initramfs and the main system are often different. If the deserialized
-         *   state is DEVICE_PLUGGED, we need to downgrade it to DEVICE_TENTATIVE (or DEVICE_DEAD if nobody
-         *   sees the device). Unlike the other starting mode, Manager.honor_device_enumeration == false
-         *   (maybe, it is better to rename the flag) when device_coldplug() and device_catchup() are called.
-         *   Hence, let's conditionalize the operations by using the flag. After switch-root, systemd-udevd
-         *   will (re-)process all devices, and the Device.found and Device.state will be adjusted.
+         *   Unlike the other starting mode, Manager.honor_device_enumeration == false (maybe, it is better
+         *   to rename the flag) when device_coldplug() and device_catchup() are called. Hence, let's
+         *   conditionalize the operations by using the flag. After switch-root, systemd-udevd will
+         *   (re-)process all devices, and the Device.found and Device.state will be adjusted.
          *
          * - On reload or reexecute, we can trust enumerated_found, deserialized_found, and deserialized_state.
          *   Of course, deserialized parameters may be outdated, but the unit state can be adjusted later by
          *   device_catchup() or uevents. */
 
-        if (!m->honor_device_enumeration && !MANAGER_IS_USER(m)) {
-                found &= ~DEVICE_FOUND_UDEV; /* ignore DEVICE_FOUND_UDEV bit */
-                if (state == DEVICE_PLUGGED)
-                        state = DEVICE_TENTATIVE; /* downgrade state */
-                if (found == DEVICE_NOT_FOUND)
-                        state = DEVICE_DEAD; /* If nobody sees the device, downgrade more */
-        }
+        /* First, let's put the deserialized state and found mask into effect, if we have it. */
+        if (d->deserialized_state < 0)
+                return 0;
 
-        if (d->found == found && d->state == state)
+        if (d->found == d->deserialized_found && d->state == d->deserialized_state)
                 return 0;
 
-        d->found = found;
-        device_set_state(d, state);
+        d->found = d->deserialized_found;
+        device_set_state(d, d->deserialized_state);
```

Comment:
> You'll call `device_set_state(DEVICE_PLUGGED)` here if `deserialized_state == DEVICE_PLUGGED`. Note that you also set `DEVICE_FOUND_UDEV` in `d->found`.

Right.

> This will cause #23208 to occur again. 

Ah, maybe. I will re-consider the issue.

Reply:
More seriously, I noticed that this re-introduces #12953. I will update this soon.

Reply:
I am still debugging. But you're generating code faster than me catching up :grinning: